chapter/methodology.md

```diff
@@ -0,0 +1,1166 @@
+# Design implementation of NLS
+
+This chapter guides through the components of the Nickel Language Server (NLS) as well as the implementation details of the source code analysis and information querying.
+Aiming for an abstract interface, NLS defines its own data structure underpinning all higher level LSP interactions.
+[Section @sec:linearization] will introduce this linearization data structure and explain how NLS bridges the gap from the explicitly handled Nickel AST.
+Finally, the implementation of current LSP features is discussed in [@sec:lsp-server-implementation].
+
+## Key Objectives
+
+The following points are considered key objectives of this thesis implemented in particular for the Nickel Language Server.
+
+### Performance
+
+The usefulness of a language server correlates with its performance.
+It may cause stutters in the editor, or prompt users to wait for responses when upon issuing LSP commands.
+Different studies suggest that interruptions are detrimentatl to programmers productivity [@interruption-1, @interruption-2]. The more often and longer a task is interrupted the higher the frustration.
+Hence, as called for in RQ.1 (cf. [@sec:research-questions]), a main criterion for the language server is its performance.
+
+Speaking of language servers there are two tasks that require processing, and could potentially cause interruptions.
+
+Upon source code changes, a language server may reprocess the code to gather general information, and provide diagnostics.
+Since, for this the LSP uses notifications, and language servers generally run as separate processes, delays in processing may not directly affect the programmer.
+However, depending on the implementation of the server, multiple changes may queue up preventing the timely response to other requests or delaying diagnostics.
+
+The JSON-RPC protocol underlying the LSP, is a synchronous protocol.
+Each request requires that the server responded to the previous request before processing can begin.
+Moreover, the order of requests has to be maintained.
+Since many requests are issued implicitly by the editor, e.g., hover requests, there is a risk of request queuing which could delay the processing of explicit commands.
+It is therefore important to provide nearly instantaneous replies to requests.
+
+It is to mention that the LSP defines "long running" requests, that may run in the background.
+This concept mitigates queuing but can lead to similarly bad user experience as responses appear out of order or late.
+
+### Capability
```

Comment:
Required capabilities

Reply:
Added action item to change all references